tx-pool/src/chunk_process.rs

```diff
@@ -0,0 +1,220 @@
+        let mut tx_snap = None;
+
+        if let Some(ref cached) = cached {
+            match cached {
+                CacheEntry::Completed(completed) => {
+                    let completed =
+                        TimeRelativeTransactionVerifier::new(&rtx, snapshot.as_ref(), &tx_env)
+                            .verify()
+                            .map(|_| *completed)
+                            .map_err(Reject::Verification)?;
+
+                    let entry = TxEntry::new(rtx, completed.cycles, fee, tx_size);
+                    self.handle.block_on(
+                        self.service
+                            .submit_entry(completed, tip_hash, entry, status),
+                    )?;
+                    self.handle
+                        .block_on(self.service.after_process(tx, remote, &Ok(completed)));
+                    self.handle.block_on(self.remove_front());
+                    return Ok(false);
+                }
+                CacheEntry::Suspended(suspended) => {
+                    tx_snap = Some(Arc::clone(&suspended.snap));
+                }
+            }
+        }
+
+        let data_loader = snapshot.as_data_provider();
+        let fee =
+            ContextualWithoutScriptTransactionVerifier::new(&rtx, consensus, &data_loader, &tx_env)
+                .verify()
+                .map_err(Reject::Verification)?;
+        let script_verifier = ScriptVerifier::new(&rtx, consensus, &data_loader, &tx_env);
+
+        let max_cycles = if let Some((declared_cycle, _peer)) = remote {
+            declared_cycle
+        } else {
+            consensus.max_block_cycles()
+        };
+        let current_cycles = tx_snap.as_ref().map(|s| s.current_cycles).unwrap_or(0);
+
+        let completed: Completed = loop {
+            // Should get here until there is no command, otherwise there maybe have a very large delay
+            while let Ok(cmd) = self.recv.try_recv() {
+                match cmd {
+                    Command::Suspend => {
+                        if let Some(snap) = tx_snap {
+                            let txs_verify_cache = Arc::clone(&self.service.txs_verify_cache);
+                            self.handle.block_on(async move {
+                                let mut guard = txs_verify_cache.write().await;
+                                guard.put(tx_hash, CacheEntry::suspended(Arc::clone(&snap), fee));
+                            })
+                        }
+                        return Ok(false);
+                    }
+                    Command::Stop => {
+                        return Ok(true);
+                    }
+                    _ => (),
+                }
+            }
+
+            let step_cycle = cmp::min(MIN_STEP_CYCLE, max_cycles - current_cycles);
+
+            let ret = if let Some(ref snap) = tx_snap {
+                script_verifier.resume(snap, step_cycle)
+            } else {
+                script_verifier.resumable_verify(step_cycle)
```

Comment:
Every time you build a snapshot is almost equivalent to copy a 4mb memory. If you don’t need to verify the block, I don’t think you should use the resumable_verify function here. You only need to take a snapshot when you are sure to receive the command to verify the block.